server/authUtils.ts

```diff
@@ -0,0 +1,115 @@
+import OpenIdClient from "openid-client";
+import { Request, Response, NextFunction } from "express";
+
+import * as Config from "./config";
+
+type OboToken = {
+  accessToken: string;
+  expiresIn: number;
+};
+
+type CachedOboToken = {
+  token: OboToken;
+  expires: number;
+};
+
+type Scope = string;
+
+declare module "express-session" {
+  export interface SessionData {
+    tokenCache: { [clientId: Scope]: CachedOboToken };
+  }
+}
+
+const OBO_TOKEN_EXPIRATION_MARGIN_SECONDS = 60;
+
+const isExpired = (token: CachedOboToken) => {
+  return (
+    token.expires < Date.now() - OBO_TOKEN_EXPIRATION_MARGIN_SECONDS * 1000
+  );
+};
+
+export const ensureAuthenticated = () => {
+  return async (req: Request, res: Response, next: NextFunction) => {
+    if (req.headers["authorization"]) {
+      return next();
+    }
+    res.status(401).send("Unauthorized");
+  };
+};
+
+const retrieveToken = (req: Request) => {
+  return req.headers.authorization?.replace("Bearer ", "") as string;
+};
+
+export const getOrRefreshOnBehalfOfToken = async (
+  authClient: OpenIdClient.Client,
+  req: Request,
+  clientId: string
+) => {
+  if (req.session.tokenCache === undefined) {
+    req.session.tokenCache = {};
+  }
+
+  var cachedOboToken = req.session.tokenCache[clientId];
+  if (!cachedOboToken || isExpired(cachedOboToken)) {
+    const token = retrieveToken(req);
+    const onBehalfOfToken = await requestOnBehalfOfToken(
+      authClient,
+      token,
+      clientId
+    );
+    cachedOboToken = {
+      token: onBehalfOfToken,
+      expires: Date.now() + onBehalfOfToken.expiresIn * 1000,
+    };
+    req.session.tokenCache[clientId] = cachedOboToken;
+  }
+  return cachedOboToken.token;
+};
+
+const requestOnBehalfOfToken = async (
+  authClient: OpenIdClient.Client,
+  access_token: string,
+  clientId: string
+) => {
+  if (!access_token) {
+    throw Error(
+      "Could not get on-behalf-of token because the access_token was undefined"
+    );
+  }
+  const grantBody = {
+    assertion: access_token,
+    client_assertion_type:
+      "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+    grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
+    requested_token_use: "on_behalf_of",
+    scope: `api://${clientId}/.default`,
+  };
+  const tokenSet = await authClient.grant(grantBody);
+  return {
```

Comment:
Burde det også vært noen sjekker på at `access_token` og `expires_in` ikke er undefined her og evt kaste feil? I stedet for å gjøre `as OboToken`

Reply:
proxyOnBehalfOf-metoden i proxy.ts (som har koden som bruker dette) kaster en feil hvis det returnerte oboToken'et ikke inneholder noe token. Kanskje det er tilstrekkelig?